components/supervisor/cmd/init.go

```diff
@@ -81,61 +65,56 @@ var initCmd = &cobra.Command{
 		case <-supervisorDone:
 			// supervisor has ended - we're all done here
 			return
-		case s := <-sigSupervisor:
+		case <-sigInput:
 			// we received a terminating signal - pass on to supervisor and wait for it to finish
-			_ = runCommand.Process.Signal(s)
-			<-supervisorDone
+			timer := time.After(cfg.GetTerminationGracePeriod())
+			terminationDone := make(chan int)
+			go func() {
+				defer close(terminationDone)
+				err := process.TerminateSync(runCommand.Process.Pid, timer)
+				if err != nil {
+					log.WithError(err).Error("supervisor termination error")
+				}
+
+				// now terminate reparented processes until there are none anymore or the time is up
+				for {
+					processes, err := procfs.AllProcs()
+					if err != nil {
+						log.WithError(err).Error("Cannot obtain processes")
+					}
+					if len(processes) == 1 {
+						close(supervisorDone)
+						return
+					}
+					// send SIGTERM to all processes but ourself
+					g := new(errgroup.Group)
+					for _, proc := range processes {
+						if proc.PID == os.Getpid() {
+							continue
+						}
+						p := proc
+						g.Go(func() error {
+							return process.TerminateSync(p.PID, timer)
+						})
+					}
+					err = g.Wait()
+					if err != nil {
+						log.WithError(err).Error("terminating child processes")
+					}
+				}
+
+			}()
+			// wait for either successful termination or the timeout
+			select {
+			case <-timer:
+				// Time is up, but we give all the goroutines a bit more time to react to this.
+				time.Sleep(time.Millisecond * 500)
```

Comment:
This would leak the goroutine (and terminationDone channel) above. Does that matter?

Reply:
The process ends here.

Reply:
In that case, should it contain `os.Exit(code)` then? It would at least communicate if it terminated gracefully, or not.

Reply:
Also, I think it's worth flushing the logger at this point to guarantee any unwritten bytes are actually written.

Reply:
> In that case, should it contain os.Exit(code) then? It would at least communicate if it terminated gracefully, or not.

But it just ends regularly here. So no reason to exit with an error code. Do I miss something?

Reply:
Normally, when you do not terminate gracefully you signal that by setting a non-zero exit code to the shell. So in this case, we'd set it when the deadline is reached

Reply:
But it is not this process that doesn't terminate gracefully, but processes that were started by users during the workspace session. I don't think we should signal through error code on this level if such a process doesn't behave well

Reply:
Good point. I guess what I'm after is to be able to tell the user (or even measure it ourselves) when we exit without completing the termination - this would then signal to users that they should possibly increase the grace termination period (in their config) but in general help them understand if the stop task did complete or not.

Reply:
Yes, that's important. The plan is to write the stdout and stderr of "shutdown" tasks [[pr](https://github.com/gitpod-io/gitpod/pull/11287)] to some file in `/workspace/.gitpod`. Do you think that is good enough? We could alternatively also write a shutdown protocol to that location.

Reply:
I think I like the idea of writing a termination protocol to a file in that folder. Will look into it.